### PR TITLE
Spike/iain ideas

### PR DESCRIPTION
**Pull request by iainkirkpatrick**
Some ideas for how the module could be organised differently, highlights:
- no classes, just functions for the data fetching / manipulation
- we can still do the "airtable singleton" approach, we just have a main function exported that wraps the others in a closure with the Airtable SDK initialised
- but now we don't have to if we don't want to, can just pass an API key etc to a specific function if we want
- using Vitest for testing, I've started using it in Bamstack / others instead of Jest (faster, seems to be becoming the community standard, Jest-compatible API)
- couldn't get Vitest working properly with PNPM and Airtable (something strange?) so went back to Yarn, I don't think it's a big deal for now, we probably move everything to PNPM in the future but especially for a library Yarn should be fine (speed and space advantages are less useful for us with a lib)
- uses Mock Service Worker to mock the network request / response from Airtable so we can test without needing to actually hit any real tables

If we were keen on these changes, we'd basically just need to fill out the rest of the functionality we want - i.e. CRUD functionality.

I haven't included ts-results for now, but that could be easily included, would obviously just change the shape of returns from each function.

@danalexilewis keen for your thoughts 😄

### CODE REVIEW

File: __tests__/constants.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export const API_KEY = "NOT_A_REAL_API_KEY"
+export const BASE_ID = "NOT_A_REAL_BASE_ID"
+export const TABLE_ID = "NOT_A_REAL_TABLE_ID"
```
Comment (danalexilewis) [inline __tests__/constants.ts]: After watching the video that @hikurangi linked to in [slack here]( https://bamboo-creative-nz.slack.com/archives/C01SJJM68J2/p1692232822781759)

I tried to use [enums for errors/constants](https://github.com/bamboocreativenz/zod-airtable/blob/a0688dddcfeba259a90c0c6007e6e846ae56aea2/errorTypes.ts#a0688dd) and really liked it.

I think it has a couple of benefits over this string literal approach:

 1. You have auto-complete when pulling in the enum.
 2. Its a more refined type then string to be used in function arguments
 
 Obviously, it doesn't really matter here for the tests, but I think its a pattern we should move to generally. Curious about your thoughts on this @iainkirkpatrick (and anyone else)

File: src/airtable.ts [hunk 2]
```diff
@@ -1,77 +1,52 @@
...
+				schema,
+				baseId,
+				query
+			}: Omit<GetRecordsWithAirtable<T>, 'airtable'>) => getRecords({
+				airtable,
+				tableId,
+				schema,
+				baseId,
+				query
 			})
```
Comment (danalexilewis) [inline src/airtable.ts]: Sorry this is a bit stream of consciousey

TLDR: I think we can just be opinionated here and say that if you have chosen to use the module in singleton mode you have to provide a `defaultBaseId`. Ideally, I'd also say the `schema` is required as well. Its how I personally envisioned using a Singleton in AFRA which is my key mental model for what is needed out of this library. If you don't want to provide these to the Singleton then you should just use the functions directly.

Ramblings...
This makes sense to me as it follows `OOP` principles. ie you use a function to curry/initialise the object and then call methods on it. If you are dealing with more then 1 base at a time this means your code will be cleaner. 

eg you would instantiate 2 singletons (1 for each base) and then when using them in throughout your code its clear which one you are calling on. 

If instead you had 1 Singleton that you were passing the baseId to all the time then you would need to know which id represented which base and remember which to pass in (or use a literal string which would also be more explicit, but at that point why use the singleton.

Honestly this is one of those things where it may just come down to irrelevant Developer preference. ie @iainkirkpatrick prefers using the functions directly and @danalexilewis prefers a singleton with baseId and schema curried in using the Factory pattern. So maybe the library just supports both well and that's all good.

File: src/getRecords.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	apiKey: string
+}
+
+export function getRecords<T extends SchemaZodT> ({
+	tableId,
+	schema,
+	baseId,
+	query,
+	...otherArgs
+}: GetRecordsWithAirtable<T> | GetRecordsWithApiKey<T>) {
```
Comment (danalexilewis) [inline src/getRecords.ts]: Overall this is so much cleaner and easier to reason with then my code - tytyty

File: src/getRecords.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import { RecordZ } from "./types/record"
+import { SchemaZodT } from "./types/schema"
+import { SelectQueryParamsT } from "./types/queryParams"
+
+export interface GetRecords<T extends SchemaZodT> {
+	tableId: string
+	baseId: string
+	schema: T
+	query?: SelectQueryParamsT
+}
```
Comment (danalexilewis) [inline src/getRecords.ts]: I personally really like `zod.functions()` for defining functions and using zods for define argument types more then using typescript. 

On the Plus: 

It means you don't need a second type definition beyond the zod and you get runtime type checks on inputs as well. Which means you could use something like a `regex` for defining the shape of a `baseId` string. Which would mean an invalid `baseId` would error and the system could send that back to the user/dev. 

I also think putting this kind of code in a library is great, as similar to generics its another more complex type of programming that has a higher bar of knowledge for the devs. But that's the point of this lib. We hide the more clever stuff in here and accept the code base as being less accessible.

However:

I noticed it didn't produce great inferred types in the static code, which annoyed me a bit. Which then required extra comments and JSDoc work to make it work. But I think this is a worthwhile compromise.

File: src/getRecords.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+	return airtable.base(baseId).table(tableId).select(query).all()
+	.then((records) => {
+		return Promise.all(
+			records.map((r) => {
+				// safeParseAsync in case of async refinements or transforms in schema, see https://zod.dev/?id=parseasync
+				return recordWithSchema.safeParseAsync(r)
+			})
+		)
+	})
```
Comment (danalexilewis) [inline src/getRecords.ts]: In principle I am all good with just returning the result of `safeParse`

But this leaves us without a strategy for how we deal with errors.

A problem I faced and something that is still unresolved in AFRA is what to do when you have errors in integrations (worst were batch operations).

To better understand how things work out the box I just spent maybe 15m going into the `airtable.js` sourcecode.

This bit was really interesting
https://github.com/Airtable/airtable.js/blob/899adb414ebc789aa3b0dcfe6c19113377315564/src/base.ts#L77

 - It shows how they do retries to make sure errors are not i/o errors
 - it shows the logic of how they resolve most errors as promise rejections with objects inside of them. [ie they don't `throw` or return `Errors`](https://github.com/Airtable/airtable.js/blob/master/src/airtable_error.ts#L1). Rather they return an error-object.

Now that I think about it, I am noticing it comes back to the question previously raised of: **what is the scope of this bamboo lib**. 

Currently the code above is a true wrapper around the airtable SDK. ie whatever Errors or error-objects are just returned to the consuming code. 

On one hand this is clean and an easy level of abstraction to find. On the other I see an opportunity to do more then this and provide devs with a more standardised approach to dealing with logic around failure.

But now thinking about this we could just do it via examples in the readme or tests.

Another example of something that could be shown in a test or in the Readme is the thing I worked out for filtering via a `.flatMap` for narrowing the types of the return.

File: src/types/schema.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import { z } from "zod"
+
+import { FieldZ } from "./airtableFields"
+
+export const SchemaZ = z.record(FieldZ)
+// N.B. this is a type for a zod
+export type SchemaZodT = z.ZodType<z.infer<typeof SchemaZ>>
```
Comment (danalexilewis) [inline src/types/schema.ts:7]: Oh this is a a wayyyyyy better way of defining the TS type for Zod with a base chemae

File: src/types/airtableFields.ts [hunk 7]
```diff
@@ -42,44 +45,7 @@ export const FieldZ = z.union([
...
-	size: number
-	type: string
-	thumbnails: {
-		small: ThumbnailT
-		large: ThumbnailT
-		full: ThumbnailT
-	}
-}
-
-export type FieldT =
```
Comment (danalexilewis) [inline src/types/airtableFields.ts:72]: completely redundent because of `type SchemaZodT` 🎉

**Review by danalexilewis — APPROVED** [review]
Overall I am super happy with this direction. 

I think there are some things to chat about and refine, but at its core I think this is a great iteration/refinement of what I started.

File: __tests__/constants.ts (hunk 1 above)
Comment (iainkirkpatrick) [reply __tests__/constants.ts]: In a case like this my immediate reaction is these aren't members of an enum @danalexilewis - i.e. they aren't members of some set in the same way that different types of errors are members of an error set. They're just specific constants that get used at specific times. I haven't watched the video recently though, maybe i'll jump on that this afternoon if I need a break from code :)

File: src/airtable.ts (hunk 2 above)
Comment (iainkirkpatrick) [reply src/airtable.ts]: Yeah totally - part of what I wanted to experiment with a bit was whether it would work to just allow for both approaches (i.e. singleton and/or assorted functions)? I think it probably can, the typings get a bit intense but if it's in a library and done then we don't have to worry about it as a consumer.

I also was trying to account for that use case where you may have multiple bases in an app that you need to call on - I can't recall us ever having needed that, at least in an app I've worked on, but it's not too hard I think to imagine a scenario where we might. That's why `defaultBaseId` is there as an option for the singleton approach - trying to recognise that most of the time you probably will want that, but you aren't forced to establish multiple singletons if you have multiple bases. Not that there is anything super wrong with multiple singletons IMO, it's just a little contrary to the basic idea of a singleton, and just means the Airtable SDK gets fired up multiple times which tbh isn't a big deal I don't think.

I'm not sure about requiring the `schema` though - that would restrict your usage of the zod-airtable singleton to a specific schema shape right? Which from where I'm at sounds like saying we can only query a specific table that matches the given schema? It felt to me like you'd want to pass the schema you are parsing results against for each call, as each call may be against a different table (let alone a different base). And even if against the same table, you might have different data parsing requirements for different calls - i.e. one call I want to parse the "Products" table records thoroughly, but another I'm just interested in "draft" Products and not all the fields have to be complete.

File: src/getRecords.ts (hunk 4 above)
Comment (iainkirkpatrick) [reply src/getRecords.ts]: Yeah that's a great call - I definitely default away from zod functions because I don't find them that useful in app code (not often that a function needs runtime parsing of arguments in app?). But I get the point that it makes a lot of sense here and in this kind of situation in general :)!

File: __tests__/constants.ts (hunk 1 above)
Comment (danalexilewis) [reply __tests__/constants.ts]: Yeah good point, these are more like env variables used for testing.

In which case it would be better to move them to a `.env.test` or jest `.env`? as they will exclusively be used for testing purposes? ie there is no prod deployment of the library.

File: src/airtable.ts (hunk 2 above)
Comment (danalexilewis) [reply src/airtable.ts]: I have 2 observations on this:

1. Similar to having a Singleton per base, I'm happy with a Singleton per table. It shouldn't matter how many we instantiate through the factory function. If the issue is that we are instantiating to many instances of the Airtable SDK we should solve this with dependency injection in the Singleton constructor/init args.

```js
const singletonTable1 = zodAirtable({
  apiKey: API_KEY,
  defaultBaseId: BASE_ID,
  schema: z.object({
    id: z.string(), 
    fields: z.object({fieldName: z.string()})
  })
})

const singletonTable2 = zodAirtable({
  apiKey: API_KEY,
  defaultBaseId: BASE_ID,
  schema: z.object({
    id: z.string(), 
    fields: z.object({fieldName: z.string()})
  })
})

const table1Data = await singletonTable1.getRecords()
const table2Data = await singletonTable2.getRecords()
```

2. The alternative would be to pass in a more complex `zod schema` that represented all the table in the base you wanted to interact with. We could then do some magic under the hood to enable something like this:

```js
const singleton = zodAirtable({
  apiKey: API_KEY,
  defaultBaseId: BASE_ID,
  schema: z.object({
    table1: z.object({id: z.string(), fields: z.object({fieldName: z.string()})}),
    table2: z.object({id: z.string(), fields: z.object({fieldName: z.string()})})
  })
})

const table1Data = await singleton.table1.getRecords()
const table2Data = await singleton.table2.getRecords()

// instead of current dx in test

const singleton = zodAirtable({
  apiKey: API_KEY,
  defaultBaseId: BASE_ID
})

const table1Data = await singleton.getRecords({
  tableId: TABLE1_ID,
  schema: z.object({
    Name: z.string(),
  })
})

const table2Data = await singleton.getRecords({
  tableId: TABLE2_ID,
  schema: z.object({
    Name: z.string(),
  })
})
```

Noticing what I specifically don't like about not having the tableId and Schema in the singleton setup is I then have to still dependency inject on every function I use from it. Which is at odds with the point of the pattern IMO.

File: src/getRecords.ts (hunk 4 above)
Comment (danalexilewis) [reply src/getRecords.ts]: > not often that a function needs runtime parsing of arguments in app

For generalised pure utility functions I would argue for parsing the args as the likelihood is that the function wouldn't work anyway with the wrong args being supplied. Most likely this happens during development and testing and so when it blows up on a failure it tells the developer there is something wrong with their logic before they call the function.

Hence I think it makes a lot of sense for a library of utility functions (like this) to use this approach.

File: src/createRecords.ts
Comment (danalexilewis) [inline src/createRecords.ts:1]: @iainkirkpatrick @hikurangi  Im thinking this could be changed to an upsert operation rather then having a create, update & upsert operation.

it would just involve adding some logic to check if the records have IDs in them.

then you separate out the ones with IDS and do an `airtable.update()` call and the others go into an `airtable.create()` call

I did something along these lines here: https://github.com/bamboocreativenz/afra-firebase/blob/4187577f58b8aceb4001893247e3cff7856209db/functions/src/airtableUtils/airtable.ts#L236

What do you all think about just standardising on upsert instead of the other 3?

File: src/createRecords.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+	return Promise.all(
+		batches.map((batch) => {
+			return airtable.base(baseId).table(tableId).create(batch)
+		})
+	).then((results) => {
+		return {
+			data: results,
+			failedValidation,
+		}
+	})
```
Comment (danalexilewis) [inline src/createRecords.ts:63]: @iainkirkpatrick @hikurangi here is an example of the behaviour I want the function to have, but I am unsure of the return type.

i.e. the code has successfully executed with batching on x number of records and returns them including the new airtableRecordIds which may need to be persisted somewhere. It also returns the records that failed validation. Its up to the Developer then to work out what to do with these.

File: src/createRecords.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+export interface CreateRecordsWithAirtable<T extends SchemaZodT>
+	extends CreateRecords<T> {
+	airtable: AirtableSDK
+}
+export interface CreateRecordsWithApiKey<T extends SchemaZodT>
+	extends CreateRecords<T> {
+	apiKey: string
+}
+
+//TODO should we change this to an upsert operation?
```
Comment (danalexilewis) [inline src/createRecords.ts:23]: @iainkirkpatrick @hikurangi whats your opinion on just having upsert rather then Create and Update functions?

File: src/getRecords.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+export function getRecords<T extends SchemaZodT>({
+	tableId,
+	baseId,
+	schema,
+	...otherArgs
+}:
+	| GetRecordsWithIdsAndAirtable<T>
+	| GetRecordsWithQueryAndAirtable<T>
+	| GetRecordsWithIdsAndApiKey<T>
+	| GetRecordsWithQueryAndApiKey<T>) {
```
Comment (danalexilewis) [inline src/getRecords.ts:58]: @iainkirkpatrick I'm not thrilled with this code, if you know of a better way to create a discriminated union with 2 points of difference with interfaces I'd be interested :)

File: src/getRecords.ts [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+	extends GetRecordsWithIds<T> {
+	apiKey: string
+}
+
+export interface GetRecordsWithQueryAndApiKey<T extends SchemaZodT>
+	extends GetRecordsWithQuery<T> {
+	apiKey: string
+}
+
+//TODO Add jsdoc as the types are not clear
```
Comment (danalexilewis) [inline src/getRecords.ts:23]: Similar to my earlier iteration, I am noticing the inferred intellisense pop over for this function is now very hard to read.

image

So we will need to write JSDoc comments on the functions to provide better documentation.

File: src/createRecords.ts [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+export function createRecords<T extends SchemaZodT>({
```
Comment (danalexilewis) [inline src/createRecords.ts:24]: @iainkirkpatrick @hikurangi whats your opinion on just having upsert rather then Create and Update functions?

Also @iainkirkpatrick just to confirm your idea here was to return the promise NOT turn this into an async/await function. This would mean the devs consuming the lib will need to to this right?

@fractalcactus sorry I'll start pinging you in these questions as well, you have have an opinion on the above too or anywhere else for that matter :)

File: src/getRecords.ts [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+export interface GetRecordsWithQueryAndApiKey<T extends SchemaZodT>
+	extends GetRecordsWithQuery<T> {
+	apiKey: string
+}
+
+export interface GetRecordsWithAirtable<T extends SchemaZodT>
+	extends GetRecordsWithIdsAndAirtable<T>,
+		GetRecordsWithQueryAndAirtable<T>,
+		GetRecordsWithIdsAndApiKey<T>,
+		GetRecordsWithQueryAndApiKey<T> {}
```
Comment (danalexilewis) [inline src/getRecords.ts]: @iainkirkpatrick I'm not thrilled with this code, if you know of a better way to create a discriminated union with 2 points of difference with interfaces I'd be interested :)

File: src/airtable.ts (hunk 2 above)
Comment (danalexilewis) [reply src/airtable.ts]: @iainkirkpatrick I've pushed a refactored version along the lines of what I wrote above.

TLDR: its simpler code that is easier to reason with and is what I was aiming for for the Singleton approach. ie setup multiple instances of zodAirtable in a single codebase.

File: src/getRecords.ts (hunk 3 above)
Comment (danalexilewis) [reply src/getRecords.ts]: So this also got refactored a bit once I started trying to implement getRecordsById into the same function.

TLDR: It got really messy once I was dealing with function interfaces for the combinations of:

 - AirtableSDK vs string
 - RecordIds vs Optional Query
 
 Breaking out the getRecordsById into its own function made everything a lot cleaner.
 
 Also noticing I really felt I was fighting against Typescript here. Without using classes there is no out the box `overloading` utility in typescript. AND trying to make discriminated types work was making this function and its types really shit. 
 
 My take away is that if we are not going to use classes with the typescript overloading method signature system then we don't want to try and write functions that have functionality determined by a discriminated union of inputs. ie what this ended up being.
 
 Heres a link to the code before I decided to break it out and it wasn't working with the Singleton approach at this point.
 
 https://github.com/bamboocreativenz/zod-airtable/blob/5df0493277aabe02f2eb187bb2272b6ad32073cc/src/getRecords.ts

File: src/getRecords.ts (hunk 13 above)
Comment (danalexilewis) [reply src/getRecords.ts]: See here for my comment on this: https://github.com/bamboocreativenz/zod-airtable/pull/1#discussion_r1306701988

**Comment by danalexilewis** [comment]
@iainkirkpatrick @hikurangi @fractalcactus I'm keen to harvest and resolving all opinions early this week and hopefully then finish fleshing out the library by the end of the week. Then we can start using this library.

I'm noticing its becoming a bit to much of an escape from other work now, so I need to find a MVP that we can live with.

File: __tests__/constants.ts (hunk 1 above)
Comment (iainkirkpatrick) [reply __tests__/constants.ts]: We could do - I think it's fine personally, they're in the `__tests__` folder, they won't be included in builds. Means developers can just run the tests without needing to get an `.env` file from somewhere, and they aren't sensitive values.

File: src/airtable.ts (hunk 2 above)
Comment (iainkirkpatrick) [reply src/airtable.ts]: @danalexilewis I've just pulled down the code to take a look - my initial thought is that the term "singleton" starts to lose a bit of meaning if we are intending that generally users should instantiate multiple "singletons", i.e. one per table they want to query.

To me this kind of just boils down to a question of where we are suggesting the table-specifying code should go. IMO it's going to be common to query multiple tables in an app that uses Airtable, so that leads me to think that specifying a table should be on a per-query basis. Querying multiple bases IMO is way less common, which is why included the `defaultBaseId` argument, but I still wanted it to be possible if you used zod-airtable without forcing the user to have to do multiple singletons.

If you require the user to pass the tableId for each query, bear in mind you can still get the kind of setup that I'm understanding you want - you just define a wrapper function in your code that closes over the zod-airtable function with the tableId already written:

```
function getRecordsFromPersonsTable ({
	apiKey,
	schema,
	query
}) {
	return getRecords({
		apiKey,
		tableId: "tbl_PERSONS",
		schema,
		query
	}) 
}
```
You can have something like that in a `utils/airtable` file etc and just import it where you need (and of course you can close over any of the other arguments too).

At the moment with the changes you've made, if you do want to query multiple tables in your app, you either have to use the `getRecords` function directly or you have to instantiate the whole zod-airtable multiple times right? I should be clear, I don't think that's the end of the world, it's not going to use up tons of memory or anything, just IMO it's an odd pattern.

I'm happy for now to just go with what works though (including just not using anything from this branch, totally fine!), AFRA is the biggest Airtable codebase we have so thinking of what suits there makes sense to me.

File: src/createRecords.ts (hunk 8 above)
Comment (iainkirkpatrick) [reply src/createRecords.ts:63]: IMO the result I'd want is the same array of records mapped, where each is now either the result of a create or an error - that means the caller can iterate over the array and decide what to do with them. Preserving the original order / index of records can sometimes be quite useful.

File: src/getRecords.ts (hunk 11 above)
Comment (iainkirkpatrick) [reply src/getRecords.ts:23]: That's interesting... I wouldn't expect it to be so intense with interfaces. I'm intrigued to look at how other library authors do this.

File: src/createRecords.ts (hunk 12 above)
Comment (iainkirkpatrick) [reply src/createRecords.ts:24]: I just default to using Promises rather than async / await - but consumers of the library can use async / await, they're interchangeable from that perspective.

Comment (iainkirkpatrick) [reply src/createRecords.ts:24]: i.e. as you have done in `airtable.test.ts`
